### DISCUSSION TITLE
WebViewPanel access to WebResourceRequested event? Need to serve generated web content

### DISCUSSION THREAD

**Discussion by Eilon** [microsoft/vscode-discussions#150] [Extension Development QnA] [answered]
I'm trying to build a VS Code extension in a WebViewPanel that displays HTML-based UI. I think WebViewPanel is the best choice because the UI is all pre-existing HTML-based content.

To implement this in various webviews, I typically need a few APIs:

1. Send/Receive WebMessage (works great in VS Code WebViewPanel)
2. Set initial static HTML content (works great too)
3. Handle requests for additional generated content/HTML (e.g. content references by `<img>`, `<link>`, `<script>`, etc. These files are not available statically, so I need to be able to write code to handle requests for those resources and get their content

It's (3) that I can't figure out how to do in a VS Code WebViewPanel.

In most WebViews in various frameworks there is access to an event such as `WebResourceRequested`, where code can handle that event and return a custom web response with whatever content it wants.

Is there a way in a VS Code WebViewPanel to return custom content for an arbitrary URL?

**Comment by burkeholland** [comment, marked as answer]
No such event like `WebResourceRequested` exists in VS Code Webviews. But you should be able to setup a service worker to act like a server for your content. Have a look here at this write-up...

https://blog.mattbierner.com/vscode-webview-web-learnings/

**Reply by Eilon** [reply]
Yeah having an actual web server is an alternative, but far more complex solution, because you have to manage a whole other lifetime. In the case I had the code I want to run ultimately calls into a .NET app, so that means I spin up ASP.NET Core to handle requests. It's very doable, but involved a lot of jumping through hoops, so not ideal.

**Reply by r3m0t** [reply]
It's not an actual web server, it's a service worker that acts as a server in the browser.
